### PR TITLE
Add basic solve dispatch with NLsolve

### PR DESCRIPTION
**Pull request by utkarsh530**
Hi,

I have added basic `NLsolve.jl` dispatch. Please also verify if the algorithm names are alright or not.
@ChrisRackauckas

### CODE REVIEW

File: src/algorithms.jl [hunk 1]
```diff
@@ -0,0 +1,9 @@
+abstract type SciMLNLSolveAlgorithm end
+
+struct JuliaNLsolve <: SciMLNLSolveAlgorithm
```
Comment (ChrisRackauckas) [inline src/algorithms.jl:3]:
```suggestion
struct NLSolveJL<: SciMLNLSolveAlgorithm

```

File: src/algorithms.jl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+abstract type SciMLNLSolveAlgorithm end
+
+struct JuliaNLsolve <: SciMLNLSolveAlgorithm
+
```
Comment (ChrisRackauckas) [inline src/algorithms.jl:4]: It should have method choices: https://github.com/JuliaNLSolvers/NLsolve.jl#fine-tunings

File: src/solve.jl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        else # Then it's an in-place function on an abstract array
+            f! = (du, u) -> (prob.f(reshape(du, sizeu), reshape(u, sizeu), p);
+            du = vec(du);
+            0)
+        end
+    end
+    u = zero(u0)
+    resid = similar(u)
+
+    if typeof(alg) <: JuliaNLsolve
```
Comment (ChrisRackauckas) [inline src/solve.jl:54]: Make sure to add the sparsity handling: https://github.com/JuliaNLSolvers/NLsolve.jl#if-the-jacobian-is-sparse

File: src/algorithms.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply src/algorithms.jl:4]: And an autodiff choice: https://github.com/JuliaNLSolvers/NLsolve.jl#automatic-differentiation

File: src/solve.jl (hunk 3 above)
Comment (utkarsh530) [reply src/solve.jl]: I guess you mean when the user specifies `j!`, call nlsolve via this:
```
df = OnceDifferentiable(f!, j!, x0, F0, J0)
nlsolve(df, initial_x)
```
Does it automatically handle sparsity? Or else do I need to do something?
Also how to handle `jac_prototype`, if user passes it in the `ODEFunction`? Or does this prototyping is computed by NLSolve, so maybe don't parse this parameter?

File: src/algorithms.jl (hunk 2 above)
Comment (utkarsh530) [reply src/algorithms.jl]: Done

File: src/algorithms.jl (hunk 1 above)
Comment (utkarsh530) [reply src/algorithms.jl]: Done

**Pull request merged**